File: src/main/scala/kaimere/tools/random/GoRN.scala
package kaimere.tools.random

import kaimere.tools.random.distributions._

object GoRN
  extends DiscreteUniform
    with ContinuousUniform
    with Normal {

  private val core = new scala.util.Random()

  def resetCore(seed: Long): Unit = core.setSeed(seed)

  override def getDiscreteUniform(min: Int, max: Int): Int = min + core.nextInt(max - min + 1)

  override def getContinuousUniform(min: Double, max: Double): Double = min + core.nextDouble() * (max - min)

  override def getNormal(mu: Double, sigma: Double): Double = {

    var x = getContinuousUniform(-1.0, 1.0)
    var y = getContinuousUniform(-1.0, 1.0)
    var s = x * x + y * y

    while (s > 1) {
      x = getContinuousUniform(-1.0, 1.0)
      y = getContinuousUniform(-1.0, 1.0)
      s = x * x + y * y
    }

    val z = x * math.sqrt(-2 * math.log(s) / s)
    mu + sigma * z

  }

  def getFromSeries[T](data: Seq[T], n: Int, withReturn: Boolean): Seq[T] =
    withReturn match {
      case true => Seq.fill(n)(getDiscreteUniform(0, data.size - 1)).map(x => data(x))
      case false => data.map(x => (x, getContinuousUniform(0.0, 1.0))).sortBy(_._2).map(_._1).take(n)
    }

}

File: Cores/DotNet/Random/GoRN.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestPlatform.TestHost;
using OSOL.Extremum.Cores.DotNet.Random.Distributions;

namespace OSOL.Extremum.Cores.DotNet.Random
{
    public class GoRN : IDiscreteUniform, IContinuousUniform, INormal
    {
        private readonly System.Random _seed;

        public GoRN(int seed)
        {
            this._seed = new System.Random(seed);
        }

        public GoRN()
        {
            this._seed = new System.Random();
        }

        public int GetDiscreteUniform(int min, int max) => _seed.Next(min, max + 1);

        public double GetContinuousUniform(double min, double max) => min + _seed.NextDouble() * (max - min);

        public double GetNormal(double mu, double sigma)
        {
            double x = GetContinuousUniform(-1.0, 1.0);
            double y = GetContinuousUniform(-1.0, 1.0);
            double s = x * x + y * y;

            while (s > 1)
            {
                x = GetContinuousUniform(-1.0, 1.0);
                y = GetContinuousUniform(-1.0, 1.0);
                s = x * x + y * y;
            }

            double z = x * Math.Sqrt(-2 * Math.Log(s) / s);
            return mu + sigma * z;
        }

        public List<T> GetFromSeries<T>(List<T> data, int n, bool withReturn)
        {
            var seq = new List<T>();
            int size = data.Count();
            if (withReturn)
            {
                for (int i = 0; i < n; ++i)
                {
                    seq.Add(data[GetDiscreteUniform(0, size - 1)]);
                }
            }
            else
            {
                seq.AddRange(
                    data
                        .Select(v => Tuple.Create(v, GetContinuousUniform(0.0, 1.0)))
                        .OrderBy(_ => _.Item2)
                        .Select(_ => _.Item1)
                        .Take(n));
            }

            return seq;
        }

    }

}